src/clients/c++/inference_profiler.cc

```diff
@@ -0,0 +1,452 @@
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "src/clients/c++/inference_profiler.h"
+
+
+namespace perfclient {
+
+nic::Error
+InferenceProfiler::Create(
+    const bool verbose, const bool profile, const double stable_offset,
+    const uint64_t measurement_window_ms, const size_t max_measurement_count,
+    const int64_t percentile, std::shared_ptr<ContextFactory>& factory,
+    std::unique_ptr<LoadManager> manager,
+    std::unique_ptr<InferenceProfiler>* profiler)
+{
+  std::unique_ptr<nic::ProfileContext> profile_ctx;
+  std::unique_ptr<nic::ServerStatusContext> status_ctx;
+  RETURN_IF_ERROR(factory->CreateProfileContext(&profile_ctx));
+  RETURN_IF_ERROR(factory->CreateServerStatusContext(&status_ctx));
+
+  profiler->reset(new InferenceProfiler(
+      verbose, profile, stable_offset, measurement_window_ms,
+      max_measurement_count, (percentile != -1), percentile,
+      factory->IsSequenceModel(), factory->ModelName(), factory->ModelVersion(),
+      std::move(profile_ctx), std::move(status_ctx), std::move(manager)));
+  return nic::Error::Success;
+}
+
+InferenceProfiler::InferenceProfiler(
+    const bool verbose, const bool profile, const double stable_offset,
+    const int32_t measurement_window_ms, const size_t max_measurement_count,
+    const bool extra_percentile, const size_t percentile,
+    const bool on_sequence_model, const std::string& model_name,
+    const int64_t model_version,
+    std::unique_ptr<nic::ProfileContext> profile_ctx,
+    std::unique_ptr<nic::ServerStatusContext> status_ctx,
+    std::unique_ptr<LoadManager> manager)
+    : verbose_(verbose), profile_(profile),
+      measurement_window_ms_(measurement_window_ms),
+      max_measurement_count_(max_measurement_count),
+      extra_percentile_(extra_percentile), percentile_(percentile),
+      on_sequence_model_(on_sequence_model), model_name_(model_name),
+      model_version_(model_version), profile_ctx_(std::move(profile_ctx)),
+      status_ctx_(std::move(status_ctx)), manager_(std::move(manager))
+{
+  load_parameters_.stable_offset = stable_offset;
+}
+
+nic::Error
+InferenceProfiler::Profile(
+    const size_t concurrent_request_count, PerfStatus& status_summary)
+{
+  status_summary.concurrency = concurrent_request_count;
+
+  RETURN_IF_ERROR(manager_->ChangeConcurrencyLevel(concurrent_request_count));
+
+  // Start measurement
+  bool is_complete = true;
+  load_parameters_.infer_per_sec.clear();
+  load_parameters_.latencies.clear();
+  load_parameters_.avg_ips = 0;
+  load_parameters_.avg_latency = 0;
```

Comment:
As load_parameters_ is used only in this function and it will be reset between calls. You can just make it a local variable instead.

Reply:
have a look at above response.